### ISSUE TITLE
Thread executor main loop is easily broken by unchecked exceptions

### ISSUE THREAD

**Issue by t-foretnik** [Refinitiv/Real-Time-SDK#198]
Unchecked exceptions can traverse all the way up to the ThreadPoolExecutor thereby breaking the main loop and render the OmmBaseImpl instance inoperable.

Clients can take care of _NOT_ passing on unchecked execptions, but the referenced [test case](https://github.com/t-foretnik/Real-Time-SDK/tree/master/Java/Eta/TestCase) shows that the library code itself will create situtations where unchecked exceptions are thrown. i.e. _NullPointerException_: 
```
Exception in thread "pool-2-thread-1" java.lang.NullPointerException
	at com.refinitiv.ema.access.ItemCallbackClientConsumer.notifyOnAllMsg(ItemCallbackClient.java:2561)
	at com.refinitiv.ema.access.ItemCallbackClient.processRefreshMsg(ItemCallbackClient.java:1788)
	at com.refinitiv.ema.access.ItemCallbackClient.defaultMsgCallback(ItemCallbackClient.java:1626)
	at com.refinitiv.eta.valueadd.reactor.Reactor.sendDefaultMsgCallback(Reactor.java:2084)
	at com.refinitiv.eta.valueadd.reactor.Reactor.sendAndHandleDefaultMsgCallback(Reactor.java:2099)
	at com.refinitiv.eta.valueadd.reactor.WlItemHandler.callbackUser(WlItemHandler.java:2943)
	at com.refinitiv.eta.valueadd.reactor.WlItemHandler.readRefreshMsg(WlItemHandler.java:2174)
	at com.refinitiv.eta.valueadd.reactor.WlItemHandler.readMsg(WlItemHandler.java:1908)
	at com.refinitiv.eta.valueadd.reactor.Watchlist.readMsg(Watchlist.java:300)
	at com.refinitiv.eta.valueadd.reactor.Reactor.processRwfMessage(Reactor.java:3143)
	at com.refinitiv.eta.valueadd.reactor.Reactor.performChannelRead(Reactor.java:3465)
	at com.refinitiv.eta.valueadd.reactor.Reactor.dispatchChannel(Reactor.java:2672)
	at com.refinitiv.eta.valueadd.reactor.ReactorChannel.dispatch(ReactorChannel.java:610)
	at com.refinitiv.ema.access.OmmBaseImpl.rsslReactorDispatchLoop(OmmBaseImpl.java:1543)
	at com.refinitiv.ema.access.OmmBaseImpl.run(OmmBaseImpl.java:1685)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
```
I created a clone of the Real-Time-SDK and added a test case sub-project to ETA. You can check it out [here](https://github.com/t-foretnik/Real-Time-SDK/tree/master/Java/Eta/TestCase)

**Comment by fogol** [comment]
@t-foretnik Thank you for bringing this to our attention. We will investigate the issue. It will be fixed in further releases.

**Comment by vlevendel** [comment]
@t-foretnik  In your testing, what is the exact sequence of events that leads to the nullpointer exception? We do not intend to add a catchall on notifyOnAllMsg because that will impact performance. However, we'd like to address any specific issues that cause NPEs.

**Comment by t-foretnik** [comment]
@vlevendel Did you read the description of the [TestCase](https://github.com/t-foretnik/Real-Time-SDK/tree/master/Java/Eta/TestCase) sub-project I added to the fork? The actual problem we encountered is outlined by [ChainConstituentDataTest.java](https://github.com/t-foretnik/Real-Time-SDK/blob/master/Java/Eta/TestCase/src/test/java/com/oraise/ChainConstituentDataTest.java). The JavaDoc on the class pretty clearly describes the issue and the test reproduces it reliably.

**Comment by NarimanAB** [comment]
Latest SDK 3.7.0.0. I periodically get the same Nullpointer exception:
`Exception in thread "pool-5-thread-1" java.lang.NullPointerException: Cannot invoke "com.refinitiv.ema.access.OmmConsumerClient.onAllMsg(com.refinitiv.ema.access.Msg, com.refinitiv.ema.access.OmmConsumerEvent)" because the return value of "com.refinitiv.ema.access.Item.client()" is null
        at com.refinitiv.ema.access.ItemCallbackClientConsumer.notifyOnAllMsg(ItemCallbackClient.java:2580)
        at com.refinitiv.ema.access.ItemCallbackClient.statusEventCallback(ItemCallbackClient.java:1185)
        at com.refinitiv.eta.valueadd.reactor.Reactor.sendTunnelStreamStatusEventCallback(Reactor.java:3044)
        at com.refinitiv.eta.valueadd.reactor.Reactor.sendAndHandleTunnelStreamStatusEventCallback(Reactor.java:3056)
        at com.refinitiv.eta.valueadd.reactor.Reactor.handleTunnelStreamMsg(Reactor.java:8276)
        at com.refinitiv.eta.valueadd.reactor.Reactor.sendAndHandleDefaultMsgCallback(Reactor.java:2925)
        at com.refinitiv.eta.valueadd.reactor.WlItemHandler.callbackUser(WlItemHandler.java:3029)
        at com.refinitiv.eta.valueadd.reactor.WlItemHandler.readMsg(WlItemHandler.java:2002)
        at com.refinitiv.eta.valueadd.reactor.WlItemHandler.serviceDeleted(WlItemHandler.java:2893)
        at com.refinitiv.eta.valueadd.reactor.WlServiceCache.clearCache(WlServiceCache.java:346)
        at com.refinitiv.eta.valueadd.reactor.WlDirectoryHandler.deleteAllServices(WlDirectoryHandler.java:762)
        at com.refinitiv.eta.valueadd.reactor.WlDirectoryHandler.deleteAllServices(WlDirectoryHandler.java:1362)
        at com.refinitiv.eta.valueadd.reactor.Watchlist.channelDown(Watchlist.java:516)
        at com.refinitiv.eta.valueadd.reactor.Reactor.sendAndHandleChannelEventCallback(Reactor.java:2721)
        at com.refinitiv.eta.valueadd.reactor.Reactor.performChannelRead(Reactor.java:5053)
        at com.refinitiv.eta.valueadd.reactor.Reactor.dispatchAll(Reactor.java:7705)
        at com.refinitiv.ema.access.OmmBaseImpl.rsslReactorDispatchLoop(OmmBaseImpl.java:1771)
        at com.refinitiv.ema.access.OmmBaseImpl.run(OmmBaseImpl.java:1920)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
`
The error is written to the system out, despite the fact that I have properly configured logger. Another strange thing is that this error happens on Saturdays only. It has already happened at least 3 weeks in a row. Could it be a coincidence?

The exception happens at the time this error is written to the log
`[pool-5-thread-1] ConsumerClient - Status message: StatusMsg
    streamId="7"
    domain="System Domain"
    privateStream
    state="Closed, Recoverable / Suspect / None / 'socket channel is not in the active state for getBuffer'"
    name="CONTRIBUTION_TUNNEL"
    serviceId="10"
    serviceName="DDS_TRCE"
StatusMsgEnd`

**Comment by ayurov2** [comment]
@NarimanAB Thanks for the additional information- we are currently working on a fix for this issue that is due out with our Q2 release. This thread will also be updated when it is released.

**Comment by vlevendel** [comment]
Addressed with [RTSDK-2.1.1.L1](https://github.com/Refinitiv/Real-Time-SDK/tree/Real-Time-SDK-2.1.1.L1).  Please let us know if there are further issues/questions.
